src/Contract/Test/Assert.purs

```diff
@@ -0,0 +1,600 @@
+    "Unexpected " <> show mdLabel <> " metadata value" <> show expectedActual
+
+  show (UnexpectedRefScriptInOutput txOutput expectedActual) =
+    "Unexpected reference script in output "
+      <> (show txOutput <> show expectedActual)
+
+  show (UnexpectedTokenDelta addr tn expectedActual) =
+    "Unexpected token delta " <> show tn <> " at address "
+      <> (show addr <> show expectedActual)
+
+  show (MaxExUnitsExceeded maxExUnits exUnits) =
+    "ExUnits limit exceeded: spent " <> show exUnits
+      <> ", but the limit is "
+      <> show maxExUnits
+
+  show (CustomFailure msg) = msg
+  show (SkippedTest msg) = msg
+
+showTxHash :: TransactionHash -> String
+showTxHash = byteArrayToHex <<< unwrap
+
+type Label = String
+
+data Labeled (a :: Type) = Labeled a (Maybe Label)
+
+label :: forall (a :: Type). a -> Label -> Labeled a
+label x l = Labeled x (Just l)
+
+unlabel :: forall (a :: Type). Labeled a -> a
+unlabel (Labeled x _) = x
+
+noLabel :: forall (a :: Type). a -> Labeled a
+noLabel = flip Labeled Nothing
+
+instance Show a => Show (Labeled a) where
+  show (Labeled _ (Just l)) = l
+  show (Labeled x Nothing) = show x
+
+data ExpectedActual (a :: Type) = ExpectedActual a a
+
+derive instance Functor ExpectedActual
+
+instance Show a => Show (ExpectedActual a) where
+  show (ExpectedActual expected actual) =
+    " (Expected: " <> show expected <> ", Actual: " <> show actual <> ")"
+
+--------------------------------------------------------------------------------
+-- Different types of assertions, Assertion composition, Basic functions
+--------------------------------------------------------------------------------
+
+-- | An check that can run some initialization code before the `Contract` is run
+-- | and check the results afterwards. It is used to implement assertions that
+-- | require state monitoring, e.g. checking gains at address.
+type ContractCheck a =
+  ContractAssertion a
+  -> ContractAssertion (ContractAssertion a /\ ContractAssertion Unit)
+
+-- | Create a check that simply asserts something about a `Contract` result.
+-- |
+-- | If a `Contract` throws an exception, the assertion is never checked,
+-- | because the result is never computed.
+assertionToCheck
+  :: forall (a :: Type)
+   . String
+  -> (a -> ContractAssertion Unit)
+  -> ContractCheck a
+assertionToCheck description f contract = do
+  putRef /\ getRef <- tieRef
+  let
+    run = do
+      res <- contract
+      putRef res
+    finalize = do
+      getRef >>= case _ of
+        Nothing -> tellFailure $ SkippedTest description
+        Just res -> f res
+  pure $ run /\ finalize
+
+assertContract
+  :: ContractAssertionFailure
+  -> Boolean
+  -> ContractAssertion Unit
+assertContract failure cond
+  | cond = pure unit
+  | otherwise = tellFailure failure
+
+assertContractMaybe
+  :: forall (a :: Type)
+   . ContractAssertionFailure
+  -> Maybe a
+  -> ContractAssertion a
+assertContractMaybe msg =
+  maybe (liftEffect $ throw $ show msg) pure
```

Comment:
If such an assertion fails, does this mean that subsequent assertions won't be executed?

Reply:
No, the finalization parts will still be executed.